File: LAB QUIZZES/CSE221_Quiz03/_A3A/_A3A.py
import os, sys, time, random, threading, subprocess
from _A3A_Solution import solve
start = 0; finish = 0; elapsed = 0
test = 0; best = 0; score = 0; batch = 0; total = 0
timeLimit = 1; nBatch = 14; N = 0
weight = [0, 5, 5, 5, 5,   8,   8,   8,   8,    8,    8,    8,     8,     8,       8]
nTest =  [0, 1, 1, 1, 1, 100, 100,  10, 100,   10,  100,  100,    10,    10,      10]
maxN =  [ 0, 5, 5, 5, 7,  10,  20, 100, 100, 1000, 1000, 1000, 10000, 100000, 100000]
A = []; res = 0
def printCase(verdict, test, A):
    if os.path.getsize(verdict + ".txt") == 0:
        with open(verdict + ".txt", "w") as fout:
            fout.write("batch = " + str(batch) + "\ntest = " + str(test) + "\n")
            fout.write("N = " + str(len(A)) + "\n")
            fout.write("A = " + str(A) + "\n")
    print(verdict + " on Batch " + str(batch), flush=True)
def exitCase(verdict):
    printCase(verdict, test, A.copy())
    os._exit(1)
def checkSoln():
    global start, finish, elapsed, list1, list2, res
    start = time.time()
    res = solve(A.copy())
    finish = time.time()
    elapsed += finish - start
    st = []
    st.append((0, len(A) - 1))
    temp = [0] * len(A)
    while(len(st) > 0):
        a, b = st.pop()
        if(a == b):
            continue
        if(b < 0):
            b = -b
            m = (a + b) >> 1
            i = a ; j = m + 1
            while(i <= m or j <= b):
                if(i > m or j <= b and A[j] < A[i]):
                    temp[i + j - a - m - 1] = A[j]
                    j += 1
                else:
                    res -= j - m - 1
                    temp[i + j - a - m - 1] = A[i]
                    i += 1
            for i in range(a, b + 1):
                A[i] = temp[i - a]
        else:
            m = (a + b) >> 1
            st.append((a, -b))
            st.append((a, m))
            st.append((m + 1, b))
    
    if(res != 0):
        exitCase("WrongAnswer")

def processBatch():
    global test, N, A
    if batch == 1:
        test = 1
        A = [1, 2, 5, 4, 3]
        N = len(A)
        checkSoln()
        return
    if batch == 2:
        test = 1
        A = [1, 2, 3, 4, 5]
        N = len(A)
        checkSoln()
        return
    if batch == 3:
        test = 1
        A = [5, 4, 3, 2, 1]
        N = len(A)
        checkSoln()
        return
    if batch == 4:
        test = 1
        A = [6, 4, 2, 5, 7, 3, 1]
        N = len(A)
        checkSoln()
        return
    test = 1
    while test <= nTest[batch]:
        N = random.randint(1, maxN[batch])
        A = [0] * N
        temp = [0] * (N + 1)
        for i in range(N):
            A[i] = random.randint(1, N)
            temp[A[i]] += 1
        for i in range(1, N + 1):
            temp[i] += temp[i - 1]
        for i in range(N):
            A[i] = temp[A[i]]
            temp[A[i]] -= 1
        checkSoln()
        test += 1
def limitTime():
    recall = -1
    while elapsed < timeLimit and recall != elapsed:
        recall = elapsed
        time.sleep(timeLimit << 1)
    exitCase("TimeLimitExceeded")
if len(sys.argv) == 2:
    batch = int(sys.argv[1])
    random.seed(batch)
if 1 <= batch and batch <= nBatch:
    print("Running on Batch " + str(batch), flush=True)
    threading.Thread(target=limitTime, daemon=True).start()
    processBatch()
    print("Accepted on Batch " + str(batch) + " in " + f"{elapsed:.9f}" + "s", flush=True)
    os._exit(0)
ID = ""
with open("EnterIDandLanguage.txt") as fin:
    ID = fin.readline().split()[0]
if os.path.exists("_A3A_" + ID + ".py"):
    with open("_A3A_" + ID + ".py") as fin:
        best = int(fin.readline().split()[2])
open("WrongAnswer.txt", "w").close()
open("TimeLimitExceeded.txt", "w").close()
batch = 1
while batch <= nBatch:
    process = subprocess.run("pypy _A3A.py " + str(batch), capture_output=True, text=True)
    if process.stdout:
        print(process.stdout, end="")
    if process.stderr:
        print(process.stderr, end="")
    if process.returncode == 0:
        score += weight[batch]
    total += weight[batch]
    batch += 1
if best <= score:
    with open("_A3A_" + ID + ".py", "w") as fout:
        fout.write("# " + ID + " " + str(score) + "\n")
    subprocess.run("cmd /c echo # %COMPUTERNAME% %USERNAME%>>_A3A_" + ID + ".py")
    with open("_A3A_" + ID + ".py", "a") as fout:
        with open("_A3A_Solution.py") as code:
            fout.write(code.read())
print("Tentative score = " + str(score / total) + "/1", flush=True)
os._exit(0)

File: LAB QUIZZES/CSE221_Quiz03/_A3A/_A3A.java
import java.util.*;
import java.io.*;

public class _A3A {
    public static Random rng = new Random();
    public static long start, finish, elapsed = 0;
    public static int test = 0, best = 0, score = 0, batch = 0, total = 0;
    public static int timeLimit = 1, nBatch = 14, N = 0;
    public static int weight[] = { 0, 5, 5, 5, 5,   8,   8,   8,   8,    8,    8,    8,     8,     8,       8};
    public static int nTest[] =  { 0, 1, 1, 1, 1, 100, 100,  10, 100,   10,  100,  100,    10,    10,      10};
    public static int maxN[] =   { 0, 5, 5, 5, 7,  10,  20, 100, 100, 1000, 1000, 1000, 10000, 100000, 100000};
    public static ArrayList<Integer> A = new ArrayList<>();
    public static Long res = 0L;

    public static void printCase(String verdict, int test, ArrayList<Integer> A)
            throws Exception {
        if (new File(verdict + ".txt").length() == 0) {
            BufferedWriter fout = new BufferedWriter(new FileWriter(verdict + ".txt"));
            fout.write("batch = " + batch + "\ntest = " + test + "\n");
            fout.write("N = " + A.size() + "\n");
            fout.write("A = " + A + "\n");
            fout.close();
        }
        System.out.print(verdict + " on Batch " + batch + "\n");
    }

    public static void exitCase(String verdict) throws Exception {
        printCase(verdict, test, new ArrayList<>(A));
        System.exit(1);
    }

    public static void checkSoln() throws Exception {
        start = System.nanoTime();
        res = _A3A_Solution.solve(new ArrayList<>(A));
        finish = System.nanoTime();
        elapsed += finish - start;
        if (1000000000L * timeLimit < elapsed)
            exitCase("TimeLimitExceeded");
        int[] st = new int[2 * (N + 1)];
        int idx = 0;
        st[idx++] = 0;
        st[idx++] = A.size() - 1;
        ArrayList<Integer> temp = new ArrayList<>();
        while(idx > 0){
            int b = st[--idx];
            int a = st[--idx];
            if(a == b)continue;
            if(b < 0){
                b = -b;
                int m = a + b >> 1;
                int i = a, j = m + 1;
                while(i <= m || j <= b){
                    if(i > m || j <= b && A.get(j) < A.get(i))
                        temp.add(A.get(j++));
                    else{
                        res -= j - m - 1;
                        temp.add(A.get(i++));
                    }
                        
                }
                for(i = a; i <= b; i++)
                    A.set(i, temp.get(i - a));
                temp.clear();
            }else{
                int m = a + b >> 1;
                st[idx++] = a;
                st[idx++] = -b;
                st[idx++] = a;
                st[idx++] = m;
                st[idx++] = m + 1;
                st[idx++] = b;
            }
        }
        if(res != 0)
            exitCase("WrongAnswer");
    }

    public static void processBatch() throws Exception {
        if (batch == 1) {
            test = 1;
            A = new ArrayList<>(Arrays.asList(1, 2, 5, 4, 3));
            N = A.size();
            checkSoln();
            return;
        }
        if (batch == 2) {
            test = 1;
            A = new ArrayList<>(Arrays.asList(1, 2, 5, 4, 3));
            N = A.size();
            checkSoln();
            return;
        }
        if (batch == 3) {
            test = 1;
            A = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5));
            N = A.size();
            checkSoln();
            return;
        }
        if (batch == 4) {
            test = 1;
            A = new ArrayList<>(Arrays.asList(6, 4, 2, 5, 7, 3, 1));
            N = A.size();
            checkSoln();
            return;
        }
        
        for (test = 1; test <= nTest[batch]; ++test) {
            A.clear();
            N =  1 + rng.nextInt(maxN[batch]);
            int[] temp = new int[N + 1];
            for (int i = 0; i < N; ++i) {
                A.add(1 + rng.nextInt(N));
                temp[A.get(i)]++;
            }
            for(int i = 1; i <= N; i++)temp[i] += temp[i - 1];
            for(int i = 0; i < N; i++)A.set(i, temp[A.get(i)]--);
            checkSoln();
        }
    }

    public static void limitTime() {
        try {
            for (long recall = -1; elapsed < 1000000000L * timeLimit && recall != elapsed;) {
                recall = elapsed;
                Thread.sleep(1000 * timeLimit << 1);
            }
            exitCase("TimeLimitExceeded");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String args[]) throws Exception {
        if (args.length == 1)
            rng.setSeed(batch = Integer.parseInt(args[0]));
        if (1 <= batch && batch <= nBatch) {
            System.out.println("Running on Batch " + batch);
            new Thread(() -> limitTime()).start();
            processBatch();
            System.out.printf("Accepted on Batch " + batch + " in %.9fs\n", elapsed * 1e-9);
            System.exit(0);
        }
        BufferedReader fin = new BufferedReader(new FileReader("EnterIDandLanguage.txt"));
        String ID = fin.readLine().split("\\s+", 2)[0];
        fin.close();
        if (new File("_A3A_" + ID + ".java").exists()) {
            fin = new BufferedReader(new FileReader("_A3A_" + ID + ".java"));
            best = Integer.parseInt(fin.readLine().split("\\s+", 4)[2]);
            fin.close();
        }
        new FileOutputStream("WrongAnswer.txt").close();
        new FileOutputStream("TimeLimitExceeded.txt").close();
        for (batch = 1; batch <= nBatch; total += weight[batch], ++batch) {
            Process process = new ProcessBuilder("java", "_A3A", "" + batch).start();
            fin = new BufferedReader(new InputStreamReader(process.getInputStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            fin = new BufferedReader(new InputStreamReader(process.getErrorStream()));
            fin.lines().forEach(System.out::println);
            fin.close();
            if (process.waitFor() == 0)
                score += weight[batch];
        }
        if (best <= score) {
            BufferedWriter fout = new BufferedWriter(new FileWriter("_A3A_" + ID + ".java"));
            fout.write("// " + ID + " " + score + "\n");
            fout.close();
            new ProcessBuilder("cmd", "/c", "echo // %COMPUTERNAME% %USERNAME%>>_A3A_" + ID + ".java").start()
                    .waitFor();
            fout = new BufferedWriter(new FileWriter("_A3A_" + ID + ".java", true));
            fin = new BufferedReader(new FileReader("_A3A_Solution.java"));
            fout.write(fin.lines().collect(java.util.stream.Collectors.joining(System.lineSeparator())));
            fin.close();
            fout.close();
        }
        System.out.print("Tentative score = " + (score / (double) total) + "/1\n");
        System.exit(0);
    }
}
